src/remoterun/custompatchsender.ts

```diff
@@ -0,0 +1,176 @@
+"use strict";
+import { PatchSender } from "../remoterun/patchsender";
+import { XmlRpcProvider } from "../utils/xmlrpcprovider";
+import { FileController } from "../utils/filecontroller";
+import { ByteWriter } from "../utils/bytewriter";
+import { VsCodeUtils } from "../utils/vscodeutils";
+import { Constants } from "../utils/constants";
+import { Credential } from "../credentialstore/credential";
+import { BuildConfigItem } from "../remoterun/configexplorer";
+import { CvsSupportProvider } from "../remoterun/cvsprovider";
+import { CheckinInfo, MappingFileContent } from "../utils/interfaces";
+import * as path from "path";
+
+export class CustomPatchSender extends XmlRpcProvider implements PatchSender {
+
+    /**
+     * @returns true in case of success, otherwise false.
+     */
+    public async remoteRun(creds: Credential, configs: BuildConfigItem[], cvsProvider: CvsSupportProvider): Promise<boolean> {
+        //We might not have userId at the moment
+        if (!creds.userId) {
+            await this.authenticateIfRequired(creds);
+        }
+
+        const patch : Buffer = await this.preparePatch(cvsProvider);
+        const additionalArgs : string[] = [];
+        const checkInInfo : CheckinInfo = await cvsProvider.getRequiredCheckinInfo();
+        additionalArgs.push(`userId=${creds.userId}`);
+        additionalArgs.push(`description="${checkInInfo.message}"`);
+        additionalArgs.push(`commitType=0`); // commitType is remote run
+        const patchDestinationUrl : string = `${creds.serverURL}/uploadChanges.html`;
+        try {
+            const changeListId : string = await VsCodeUtils.makeRequest(Constants.POST_METHOD,
+                                                                        patchDestinationUrl,
+                                                                        creds,
+                                                                        patch,
+                                                                        additionalArgs);
+            const errors : string = await this.triggerChangeList(changeListId, configs, creds);
+        } catch (err) {
+            console.log(VsCodeUtils.formatErrorMessage(err));
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * This method uses PatchBuilder to write all required info into the patch file
+     * @param cvsProvider - CvsProvider object
+     * @return Promise<Buffer> of patch content
+     */
+    public async preparePatch(cvsProvider : CvsSupportProvider) : Promise<Buffer> {
+        const checkInInfo : CheckinInfo = await cvsProvider.getRequiredCheckinInfo();
+        const changedFilesNames : string[] = checkInInfo.fileAbsPaths;
+        if (!changedFilesNames) {
+            return;
```

Comment:
I'm going to add logs to all methods when the patchsender will finished, cause it will have lots of changes.